amgut/lib/data_access/redcap.py

```diff
@@ -0,0 +1,115 @@
+from urllib import urlencode
+from json import loads
+
+from tornado.web import HTTPError
+from tornado.httpclient import HTTPClient
+
+from amgut.lib.config_manager import AMGUT_CONFIG
+from amgut.lib.data_access.sql_connection import TRN
+
+
+def _make_request(data):
+    body = urlencode(data)
+    client = HTTPClient()
+    response = client.fetch(AMGUT_CONFIG.redcap_url, method='POST',
+                            headers=None, body=body)
+    # Response not always JSON, so can't always use loads
+    if '{"error"' in response.body:
+        raise HTTPError(400, loads(response.body)['error'])
+    return response.body
+
+
+def get_survey_url(record, instrument='ag_human_en_us'):
+    """Genereates a new survey link for a given record
+
+    Parameters
+    ----------
+    record : int
+        Redcap record (consent) to attach this survey to
+    instrument : str, optional
+        The instrument (survey) to get the URL for. Default ag_human_en_us
+
+    Returns
+    -------
+    str
+        URL for the survey
+    """
+    with TRN:
+        # Get the event ID for the survey, or 1 if first survey
+        sql = """SELECT MAX(redcap_event_id) + 1
+                 FROM ag.ag_login_surveys
+                 LEFT JOIN ag.ag_consent USING (redcap_record_id)
+                 LEFT JOIN redcap_instruments
+                    USING (redcap_instrument_id, lang)
+                 WHERE redcap_record_id = %s AND redcap_instrument_id = %s;
+        """
+        TRN.add(sql, [record, instrument])
+        event = TRN.execute_fetchlast()
+        if event is None:
+            event = 1
+        data = {
+            'token': AMGUT_CONFIG.redcap_token,
+            'content': 'surveyLink',
+            'format': 'json',
+            'instrument': 'ag_human',
+            'event': 'event_%d_arm_1' % event,
+            'record': record,
+            'returnFormat': 'json'
+        }
+        return _make_request(data)
+
+
+def log_complete(record, instrument, event):
+    """Logs a redcap survey as completed in the database
+
+    Parameters
+    ----------
+    records : list of int
+        record to log for
+    instrument : str
+        The instrument (survey) to log for
+    event : int
+        Event to log for
+    """
+    with TRN:
```

Comment:
The code in the next PR actually expects that the same redcap_event_id is used if a survey is not completed. When someone starts a survey but doesn't complete it, it is not logged in redcap nor the AG db. This function is called as part of the AG passthrough page, which is where participants are sent by redcap after their survey is logged. This allows us to not waste event-ids on unfinished surveys and still track what surveys people have taken.